### PR TITLE
added return types for "SeriesGroupBy" and "DataFrameGroupBy" 

### LINKED ISSUE #438: groupby(...).transform is untyped
**Describe the bug**
`SeriesGroupBy.transform` and `DataFrameGroupBy.transform` are currently not typed:

https://github.com/pandas-dev/pandas-stubs/blob/001c0afb3ef0dfe21e96ced87c9e455c657a50c3/pandas-stubs/core/groupby/generic.pyi#L159

Running mypy with `--disallow-untyped-calls` reports an issue with any usage of this method.

### PR DESCRIPTION
**Pull request by ramvikrams**
- [x] Closes #438 
- [x] Tests added: Please use `assert_type()` to assert the type of any return value

### CODE REVIEW

File: tests/test_frame.py [hunk 1]
```diff
@@ -2063,3 +2063,25 @@ def test_setitem_none() -> None:
...
+    ser = pd.Series(
+        [390.0, 350.0, 30.0, 20.0],
+        index=["Falcon", "Falcon", "Parrot", "Parrot"],
+        name="Max Speed",
+    )
+    grouped = df.groupby("A")[["C", "D"]]
+    grouped1 = ser.groupby(ser > 100)
+    c1 = grouped.transform(lambda x: (x - x.mean()) / x.std())
+    c2 = grouped1.transform(lambda x: x.max() - x.min())
+    check(assert_type(c1, "pd.DataFrame"), pd.DataFrame)
```
Comment (Dr-Irv) [inline tests/test_frame.py:2086]: don't need quotes here

File: pandas-stubs/core/groupby/generic.pyi [hunk 2]
```diff
@@ -61,7 +61,7 @@ class SeriesGroupBy(GroupBy, Generic[S1]):
     def agg(self, func: list[AggFuncTypeBase], *args, **kwargs) -> DataFrame: ...
     @overload
     def agg(self, func: AggFuncTypeBase, *args, **kwargs) -> Series: ...
-    def transform(self, func, *args, **kwargs): ...
+    def transform(self, func, *args, **kwargs) -> Series: ...
```
Comment (Dr-Irv) [inline pandas-stubs/core/groupby/generic.pyi:64]: can you add `func: Callable` here (and below)?

**Review by Dr-Irv — CHANGES_REQUESTED** [review]

File: pandas-stubs/core/groupby/generic.pyi (hunk 2 above)
Comment (ramvikrams) [reply pandas-stubs/core/groupby/generic.pyi]: Done sir

File: tests/test_frame.py (hunk 1 above)
Comment (ramvikrams) [reply tests/test_frame.py]: Done sir

File: pandas-stubs/core/groupby/generic.pyi (hunk 2 above)
Comment (ramvikrams) [reply pandas-stubs/core/groupby/generic.pyi]: after adding Callable line number 643 and 645 showa this error `Argument 1 to "transform" of "DataFrameGroupBy" has incompatible type "str"; expected "Callable[..., Any]"  [arg-type]` but if we add `Callable[..., str]` it also shows the same error

File: pandas-stubs/core/groupby/generic.pyi [hunk 3]
```diff
@@ -61,7 +61,7 @@ class SeriesGroupBy(GroupBy, Generic[S1]):
     def agg(self, func: list[AggFuncTypeBase], *args, **kwargs) -> DataFrame: ...
     @overload
     def agg(self, func: AggFuncTypeBase, *args, **kwargs) -> Series: ...
-    def transform(self, func, *args, **kwargs): ...
+    def transform(self, func: Callable, *args, **kwargs) -> Series: ...
```
Comment (Dr-Irv) [inline pandas-stubs/core/groupby/generic.pyi:64]: Looks like it needs to be `Callable | str` (in both of them)

Can you also create an issue in the pandas repo saying that the docs for `transform` need to indicate that the function parameter in `groupby.transform()` can be a string function name, and that an example should be included?

Comment (ramvikrams) [reply pandas-stubs/core/groupby/generic.pyi:64]: yes I'll do that

File: tests/test_frame.py [hunk 4]
```diff
@@ -2063,3 +2063,25 @@ def test_setitem_none() -> None:
...
+        [390.0, 350.0, 30.0, 20.0],
+        index=["Falcon", "Falcon", "Parrot", "Parrot"],
+        name="Max Speed",
+    )
+    grouped = df.groupby("A")[["C", "D"]]
+    grouped1 = ser.groupby(ser > 100)
+    c1 = grouped.transform(lambda x: (x - x.mean()) / x.std())
+    c2 = grouped1.transform(lambda x: x.max() - x.min())
+    check(assert_type(c1, pd.DataFrame), pd.DataFrame)
+    check(assert_type(c2, pd.Series), pd.Series)
```
Comment (Dr-Irv) [inline tests/test_frame.py:2087]: can you add tests for `grouped.transform("sum")` and `grouped1.transform("cumsum")`

**Review by Dr-Irv — CHANGES_REQUESTED** [review]

File: pandas-stubs/core/groupby/generic.pyi [hunk 5]
```diff
@@ -166,7 +166,7 @@ class DataFrameGroupBy(GroupBy):
     ) -> DataFrame: ...
     def aggregate(self, arg: AggFuncTypeFrame = ..., *args, **kwargs) -> DataFrame: ...
     agg = aggregate
-    def transform(self, func, *args, **kwargs): ...
+    def transform(self, func: Callable | str, *args, **kwargs) -> DataFrame: ...
```
Comment (gandhis1) [inline pandas-stubs/core/groupby/generic.pyi:169]: Can/should we make this a `Literal`?

From https://github.com/pandas-dev/pandas/blob/main/pandas/core/groupby/base.py#L60, the transformation kernels would be the ones applicable here:

```
transformation_kernels = Literal[
    "bfill",
    "cumcount",
    "cummax",
    "cummin",
    "cumprod",
    "cumsum",
    "diff",
    "ffill",
    "fillna",
    "ngroup",
    "pct_change",
    "rank",
    "shift",
]
```

Comment (gandhis1) [reply pandas-stubs/core/groupby/generic.pyi:169]: Is there any way to do better than `Callable`? Like what about `Callable[Concatenate[pd.DataFrame, P], pd.DataFrame]`? Will `ParamSpec` work in that context? If not, can we use a `Protocol`? I know that we can't use `P.args` on `*args` and `P.kwargs` on `**kwargs`, though that would have been a very nice feature.

I'm just thinking, if we are going from an unannotated function to an annotated function, a loose annotation doesn't really do that much good versus not having an annotation at all.

File: tests/test_frame.py (hunk 4 above)
Comment (ramvikrams) [reply tests/test_frame.py:2087]: Yes sir

File: pandas-stubs/core/groupby/generic.pyi (hunk 5 above)
Comment (ramvikrams) [reply pandas-stubs/core/groupby/generic.pyi:169]: I think it would be better the ask Dr Irv about adding it.

Comment (Dr-Irv) [reply pandas-stubs/core/groupby/generic.pyi:169]: I'd rather be incremental in terms of PR's where we could specify the `Literal` values (there are more than just the ones you listed), as well as making the `func` argument more specific.  @gandhis1 feel free to create an issue indicating that we could improve the typing on `groupby.transform` by being more specific about the allowed argument combinations.  It also would require a lot more tests.

Focus of this PR was just to get the return types right, and do a slight improvement in the `func` argument.

**Comment by ramvikrams** [comment]
> thanks @ramvikrams 

Thanks @Dr-Irv sir

**Pull request merged**